CPU manager phase 1

ConnorDoyle opened the pull request:
**What this PR does / why we need it**:
* Implements phases 1 and 2 of the [CPU manager design proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/cpu-manager.md). Implements the _none_ and _static_ cpu pinning policies.

E2e tests are in progress and will be added in a subsequent PR.
User / operator docs for `kubernetes/kubernetes.github.io` are merged.

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
TODO
```

--------
_Plan to break this into smaller PRs..._

- [x] add UpdateContainerResources method to CRI (893 LOC) #46105
- [x] cpuset helpers and unit tests (632 LOC) #51132
- [x] cpu manager interfaces (238 LOC) #51140
- [x] flag, feature gate, wiring into Kubelet, none policy and unit tests #51357
- [ ] Static policy, topology, takeByTopology and unit tests (1512 LOC) #51180
- [ ] CPU manager node e2e tests. (375 LOC) #51041
--------

Starting the following three pods in order on an 8-core single socket system:

**cpuset-visualizer-exclusive-1**

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cpuset-visualizer-exclusive-1
spec:
  containers:
  - image: quay.io/connordoyle/cpuset-visualizer
    name: cpuset-visualizer-exclusive-1
    resources:
      requests:
        cpu: 1
        memory: "256M"
      limits:
        cpu: 1
        memory: "256M"
```

**cpuset-visualizer-exclusive-3**

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cpuset-visualizer-exclusive-3
spec:
  containers:
  - image: quay.io/connordoyle/cpuset-visualizer
    name: cpuset-visualizer-exclusive-3
    resources:
      requests:
        cpu: 3
        memory: "256M"
      limits:
        cpu: 3
        memory: "256M"
```

**cpuset-visualizer-exclusive-4**

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: cpuset-visualizer-exclusive-4
spec:
  containers:
  - image: quay.io/connordoyle/cpuset-visualizer
    name: cpuset-visualizer-exclusive-4
    resources:
      requests:
        cpu: 3
        memory: "256M"
      limits:
        cpu: 3
        memory: "256M"
```

Gave the following allocations. Note that cpu `0` was not used for pods in this case, per kube-reserved and system-reserved as described in the proposal.

![topo-order](https://user-images.githubusercontent.com/379372/28298839-774afde4-6b2a-11e7-9f42-4b325ab1edf4.png)

cc @sjenning @flyingcougar @nqn @teferi @vishh

sjenning committed:
Initial cpumanager implementation.

ConnorDoyle committed:
Refactor cpu manager.

- Renamed "cpuset manager" => "cpu manager" everywhere.
- Introduced a new CPUSet type.
- Factored policy out of cpu manager.
- Factored state out of cpu manager policy.
- Lifted initial container resources settings and reconciliation
  loop into the cpu manager (for reuse in multiple policies.)
- Normalized log output (but it could use more cleanup.)

Commit:
Discover topology from cAdvisor.

- Move topo to state.
- Add cpu manager flag.
- Create a new type for PolicyName.
- Add unit tests for topology discovery.

teferi committed:
Handle errors during CPU assignment.

- Fix Info -> Infof typo.

This patch adds a check that container has been registered successfully
in the cpu manager. There is a sizeable period of time before deleted
containers are actually removed and unregistered from cpu manager.
During this time k8s may assign new guaranteed containers to the node.
In this case registering new containers may fail due to lack of CPUs
(occupied by to-be-deleted containers) and new containers would not get
correct CPU settings.
With this patch such containers would be restarted until registration is
successful.

sjenning committed:
Refactor cpumanager package.

- Remove duplicate state_mem_test.go file.
- Error msg fix and gofmt.
- Move discoverTopology to topo package.
- Rename topo package to topology.

ConnorDoyle committed:
Fix errors in unit tests related to cpumanager.

- Correct json serialization annotation.
- Add fake CPU manager for testing.

ConnorDoyle committed:
Static CPU manager policy takes CPUs in topo order

Other minor changes:

- Fix duplicate status manager initialization.
- Use cpumanager policy consts everywhere possible.
- Remove inconsequential lines from static allocation function.
- Remove unneeded TODO comments.
- Remove unused Policy() from cpumanager interface.

Commit:
Added unit tests for static policy in CPU manager.

ConnorDoyle commented on pkg/kubelet/apis/cri/v1alpha1/runtime/api.proto:
```diff
@@ -763,6 +769,15 @@ message ContainerStatusResponse {
+    // ID of the container to update.
+    string container_id = 1;
+    // Resource configuration specific to Linux containers.
+    LinuxContainerResources linux = 15;
```
Note: would make more sense for this to be `2`.

ConnorDoyle commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -248,6 +251,9 @@ func (m *kubeGenericRuntimeManager) generateLinuxContainerConfig(container *v1.C
 		cpuShares = milliCPUToShares(cpuRequest.MilliValue())
 	}
 	lc.Resources.CpuShares = cpuShares
+	// SETH set initial cpuset here maybe?
```
TODO: kill comment?

ConnorDoyle commented on pkg/kubelet/cpumanager/policy_static.go:
```diff
@@ -0,0 +1,226 @@
+		return cpuset.NewCPUSet(), fmt.Errorf("not enough cpus available to satisfy request")
+	}
+
+	// TODO(CD): Acquire CPUs topologically instead of sequentially...
```
TODO: delete these lines

ConnorDoyle commented on pkg/kubelet/kubelet.go:
```diff
@@ -1241,6 +1257,9 @@ func (kl *Kubelet) initializeModules() error {
 	// Start resource analyzer
 	kl.resourceAnalyzer.Start()
 
+	// Step 9: Start the CPU manager
```
Nit: remove Step 9 prefix

ConnorDoyle commented on pkg/kubemark/hollow_kubelet.go:
```diff
@@ -145,6 +145,7 @@ func GetHollowKubeletConfig(
 	c.VolumeStatsAggPeriod.Duration = time.Minute
 	c.CgroupRoot = ""
 	c.ContainerRuntime = kubetypes.DockerContainerRuntime
+	c.CPUManagerPolicy = "noop"
```
TODO: use const from cpumanager package instead of this raw string

ConnorDoyle commented on pkg/apis/componentconfig/v1alpha1/defaults.go:
```diff
@@ -229,6 +229,9 @@ func SetDefaults_KubeletConfiguration(obj *KubeletConfiguration) {
 		obj.ContainerRuntime = "docker"
 	}
+	if obj.CPUManagerPolicy == "" {
+		obj.CPUManagerPolicy = "noop"
```
TODO: use const from the cpumanager package instead of this raw string

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,208 @@
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+		newPolicy = NewStaticPolicy(topo)
+	default:
+		glog.Warningf("[cpumanager] Invalid policy, fallback to default policy - 'noop'")
```
Nit: interpolate `PolicyNoop` const in this log output

ConnorDoyle commented:
/assign @derekwaynecarr @vishh 
/unassign @pmorie @euank

ConnorDoyle commented:
@teferi and @nqn could you please help turn the CLA gate green? There's a wiki entry on how it works here: https://github.com/kubernetes/kubernetes/wiki/CLA-FAQ

sjenning commented:
/assign

sjenning commented:
@ConnorDoyle thanks for opening this PR :+1:   We _do_ need to start moving this forward and getting review if we are going to make 1.8

I did want to clean up the commit history on our dev branch before opening a PR.  I would like to maintain credit though for each author.  Maybe we can squash by author and rework the commit messages so that we don't have "wip" commits and whatnot.

sjenning commented:
Reviewer note: the first commit in this PR is the commit for https://github.com/kubernetes/kubernetes/pull/46105.  I think we might want to review the CRI change separately in that PR, which needs rebasing if we want to keep it open.

Thoughts?

ConnorDoyle commented:
I can squash consecutive commits by author and coalesce the messages. Wanted to do the big rebase first to stay sane; now we're caught up with master.

> On Jul 20, 2017, at 16:43, Seth Jennings <notifications@github.com> wrote:
> 
> @ConnorDoyle thanks for opening this PR 👍 We do need to start moving this forward and getting review if we are going to make 1.8
> 
> I did want to clean up the commit history on our dev branch before opening a PR. I would like to maintain credit though for each author. Maybe we can squash by author and rework the commit messages so that we don't have "wip" commits and whatnot.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Commit:
Take kube/system reserved cores into account

ConnorDoyle committed:
Added examples.

- Example pods.
- Helper scripts.
- Basic instructions.

ConnorDoyle committed:
Added OWNERS file to cpumanager.

Commit:
Improve test coverage for CPU Manager.

Commit:
Fix 'make verify' errors.

sjenning committed:
add support CPUPressure node condition

ConnorDoyle committed:
Reconcile returns success/failure lists for tests.

sjenning committed:
fix cpu reservation code

ConnorDoyle committed:
Export KUBECONFIG in cpumanager examples.

ConnorDoyle commented on cmd/kubelet/app/options/options.go:
```diff
@@ -237,6 +237,7 @@ func (c *kubeletConfiguration) addFlags(fs *pflag.FlagSet) {
 	fs.BoolVar(&c.CgroupsPerQOS, "cgroups-per-qos", c.CgroupsPerQOS, "Enable creation of QoS cgroup hierarchy, if true top level QoS and pod cgroups are created.")
 	fs.StringVar(&c.CgroupDriver, "cgroup-driver", c.CgroupDriver, "Driver that the kubelet uses to manipulate cgroups on the host.  Possible values: 'cgroupfs', 'systemd'")
 	fs.StringVar(&c.CgroupRoot, "cgroup-root", c.CgroupRoot, "Optional root cgroup to use for pods. This is handled by the container runtime on a best effort basis. Default: '', which means use the container runtime default.")
+	fs.StringVar(&c.CPUManagerPolicy, "cpu-manager-policy", c.CPUManagerPolicy, "CPU Manager policy to use. Possible values: 'noop', 'static'. Default: 'noop'")
```
Per the updated proposal, change all instances of `noop` to `none` in this patch set.

sjenning replied:
I noticed that reading the proposal against yesterday.  Thanks for making a note!

Commit:
Remove unnecessary test.
- Makes tests pass again.

ConnorDoyle committed:
Added best-effort pod to cpumanager examples.

ConnorDoyle committed:
Rename noop cpumnager policy => none.

- Also fixed errors in `make test`.

ConnorDoyle committed:
Ran update-bazel.sh.

sjenning committed:
add eviction on cpu pressure

ConnorDoyle commented:
NB: Now that @sjenning has implemented the CPU pressure node condition and eviction of no-CPU pods, this branch is feature-complete (e.g. implements phase 1 and 2 of the proposal). @balajismaniam is working on adding node e2e tests.

ConnorDoyle committed:
Ran update-bazel.sh.

ConnorDoyle commented:
@teferi could you please sign the CLA? We unfortunately won't be able to include your patch unless this is done.

teferi commented:
@ConnorDoyle Sure, will sign as soon as I get to my PC

sjenning commented:
@ConnorDoyle fyi, this work on the scheduler might impact our PR https://github.com/kubernetes/kubernetes/issues/42001

Basically they are doing away with explicit node condition filtering and moving to tainting nodes.

teferi commented:
CLA verification bump

teferi commented:
Bump to verify CLA.

rohitsardesai83 commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	// to make UpdateContainerResources() calls against the containers.
+	containerRuntime runtimeService
+
+	// podLister provides a method for listing all the pods on the node
```
podLister should be renamed to kletGetter in the comment ?

rohitsardesai83 commented on pkg/kubelet/cpuset/cpuset.go:
```diff
@@ -0,0 +1,260 @@
+			}
+			result.Add(elem)
+		} else if len(boundaries) == 2 {
+			// Handle multi-element rnages like "0-5".
```
rnages -> ranges

rohitsardesai83 commented on pkg/kubelet/kubelet.go:
```diff
@@ -618,6 +624,23 @@ func NewMainKubelet(kubeCfg *componentconfig.KubeletConfiguration,
+			killPodNow(klet.podWorkers, kubeDeps.Recorder),
+			kubeDeps.Recorder)
+		if err != nil {
+			glog.Infof("[cpumanager] failed to initialize cpu manager: %v", err)
```
glog.Infof -> glog.Errorf

ConnorDoyle committed:
Cleanup typos and log levels.

ConnorDoyle committed:
Added unit tests for takeByTopology in cpumanager

- Moved takeByTopology to a new file.
- Made whole-socket, whole-core and cpu branches deterministic.
- Fixed a bug where asking for zero cpus yielded a nonempty cpuset.

Commit:
Fixed sorting function in takeByTopology

- Added unit test for sorting
- Fix sorting function

ConnorDoyle committed:
Moved pressure condition from policy to state.

derekwaynecarr commented:
i plan to get a review of this done prior to sig-node tomorrow.

derekwaynecarr commented:
@ConnorDoyle @sjenning -- can we cut this down into a final set of commits?

derekwaynecarr commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	// is called after RegisterContainer(), which creates the containerID, and
+	// before StartContainer(), such that the cpuset is configured before
+	// the container starts
+	RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error
```
i think this reads a bit confusing.

ConnorDoyle replied:
How about:

```
// RegisterContainer is called between container creation and container start
// so that initial CPU affinity settings can be written through to the
// container runtime before the first process begins to execute.
```

derekwaynecarr replied:
+1

ConnorDoyle committed:
Refactored takeByTopology into smaller functions.

derekwaynecarr commented on pkg/kubelet/cpumanager/state/state_mem.go:
```diff
@@ -0,0 +1,82 @@
+var _ State = &stateMemory{}
+
+// NewMemoryState creates new State for keeping track of cpu/pod assignment
+func NewMemoryState() State {
```
this is a big pr, so bear with me, but is it safe that there is no locking anywhere here?

ConnorDoyle replied:
Added a note to the sub-PR the introduces the impl.

ConnorDoyle replied:
Fixed in sub-PR

derekwaynecarr commented on pkg/kubelet/cpumanager/topology/topology.go:
```diff
@@ -0,0 +1,161 @@
+	NumCores         int
+	HyperThreading   bool
+	NumSockets       int
+	CPUtopoDetails   CPUDetails
```
nit: CPUTopoDetails

ConnorDoyle replied:
Renamed to just CPUDetails

derekwaynecarr commented on pkg/kubelet/cpuset/cpuset.go:
```diff
@@ -0,0 +1,260 @@
+limitations under the License.
+*/
+
+package cpuset
```
this feels like it should be `/pkg/kubelet/util/cpuset`

vishh replied:
https://twitter.com/davecheney/status/697551191752855552?lang=en at-least moving forward.

ConnorDoyle replied:
@vishh do you have another preference on where it should live, maybe inside `pkg/kubelet/cm` or is it OK here?

ConnorDoyle replied:
Moving to /pkg/kubelet/cm per Vish's other comments.

ConnorDoyle replied:
Moved

derekwaynecarr reviewed (changes requested):
this is a huge pr.

i left a number of comments throughout, but can we break this into logical commits.  separate the scheduler pieces from the kubelet pieces as well.

ConnorDoyle commented on pkg/kubelet/cpumanager/state/state_mem.go:
```diff
@@ -0,0 +1,82 @@
+var _ State = &stateMemory{}
+
+// NewMemoryState creates new State for keeping track of cpu/pod assignment
+func NewMemoryState() State {
```
This is safe because the only writer to the state is the policy. There is synchronization within the CPU manager to ensure that the policy's callbacks are not executed concurrently.

To improve this, we could either put a big warning on the state saying that it is not thread-safe, or make it thread-safe. Which do you prefer?

derekwaynecarr replied:
i am fine with a big warning that says locking must be handled by caller.

ConnorDoyle replied:
On second look, we do need synchronization here. We already have concurrency hazard where the reconcile loop is reading `stateMemory.assignments` while the policy may be writing to the same map.

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+		// This really should only do something on the first reconcilation
+		// pass after transitioning to UnderPressure conditions, if all
+		// no-CPU pods are evicted successfully.
+		m.evictNoCPUPods()
```
Yes, for now. If different requirements come up for later policies we can revisit it. The bigger picture is that it needs to be part of some reconciliation loop because eviction can timeout. So, this avoids the static policy having to run a separate reconcile to manage evictions. One way around this in the future could be to add a `toEvict` list to the state, and use that here.

derekwaynecarr replied:
i am fine to defer that

vishh commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -885,6 +885,64 @@ func (kl *Kubelet) setNodeOODCondition(node *v1.Node) {
+	// condition.Status != v1.ConditionTrue or
+	// condition.Status != v1.ConditionFalse in the conditions below depending on whether
+	// the kubelet is under cpu pressure or not.
+	if kl.cpuManager.IsUnderCPUPressure() {
```
nit: Why is CPU manager exposed directly to kubelet? It would be great if this would become part of NodeContainerManager interface,

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+/*
```
Can we move cpuset to under `pkg/kubelet/cm`? In general, the architecture I'm hoping to converge on is one where the Container manager is responsible for all resource management. In container manager, we have one for node and pod. Node container manager deals with handling overall CPU capacity, allocatable, availability (pressure), etc. 
PodContainerManager deals with CPU allocation, NUMA, device allocation, and any other performance optimizations per pod. There needs to be shared state between NodeContainerManager and per pod PodContainerManager objects.
Kubelet Node Lifecycle management will have to understand NodeContainerManager APIs only.
Kubelet Pod lifecycle management will have to interact with PodContainerManager APIs only.

ConnorDoyle replied:
Moved and re-wired in the sub-PR.

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+		glog.Infof("[cpumanager] detected CPU topology: %v", topo)
+
+		resources := capacityProvider.GetNodeAllocatableReservation()
+		cpuResource, ok := resources[v1.ResourceCPU]
```
Can we reserve one core by default?

ConnorDoyle replied:
As mentioned in the other comment, if we do this then we avoid needing the CPU pressure condition and evictions but open the door to this situation:

* Given a system with 4 logical CPUs on Node N1.
* A user carefully creates pod P1 to obtain 4 exclusive cores.
* The scheduler sees 4 CPUs available on N1 and binds P1 there.
* The Kubelet synchronizes but is unable to run the pod due to the hard-coded shared pool low-water-mark.
* User is... baffled?

Furthermore, in a cluster where a large mix of exclusive CPU containers are present there would be no easy way to prevent this. Some fraction of guaranteed pods will fail. If I were operating such a cluster, I would work around the problem by advertising a separate resource to account exclusively-allocatable cores, which was an explicit anti-pattern the CPU manager design tries hard to avoid.

At this point the issue isn't necessarily one of technical correctness but a difference of opinion. The conflict thus far has been about which breakdown in user expectation is least bad. I think that having a pod with no CPU request get evicted from a node with the static CPU policy enabled is much more reasonable than the above scenario and matches the wider discussion that occurred during proposal review time.

vishh replied:
The situation you described to me sounds like an operator error. 
To avoid having users be baffled,
1. Cluster admins can be instructed to reserve CPU for system daemons and,
2. Kubelet can reject pods during admission that ask for more than available "exclusive" CPUs.
The approach of using node pressure condition for representing schedulable capacity seems like an anti-pattern. Thus far pressure conditions have been about usage.

We may help users by having kubelet require that `kubereserved.cpu + systemreserved.cpu > 0 millicpus` when static CPU pinning policy is enabled. 

I'm suggesting identifying ways that avoid the problem you mention without requiring much changes to the system.

> matches the wider discussion that occurred during proposal review time.

I apologize, I did recall the discussion around requiring users to set non zero cpu allocatable, but I did not recall requiring a new node condition.
This feature is non trivial. Let's not rush to adding an API just because of release pressure.
Adding a node condition is easy, but removing one is hard.

ConnorDoyle replied:
> Let's not rush to adding an API just because of release pressure. Adding a node condition is easy, but removing one is hard.

Point well taken.

> We may help users by having kubelet require that kubereserved.cpu + systemreserved.cpu > 0 millicpus when static CPU pinning policy is enabled.

This definitely crossed my mind too but I couldn't think of how it helps in practice. Let's expand on how the CPU manager would handle assignments with this strategy. How about doing the following:

* Require nonzero node+system reservation.
* Shared pool initially includes all logical CPUs, regardless of the reserved amount.
* The resource allocation math works out since < shared pool size is advertised in allocatable, meaning we can accommodate at most `shared pool size - 1` worth of exclusive-cpu containers.

This SGTM, would like to get @sjenning input here too since he implemented the condition and eviction and may have more insight on missing edge cases.

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
+	m.Lock()
```
Lock is being held while invoking a remote call (UpdateContainerResources). Is this a good idea?

ConnorDoyle replied:
Good catch; it should be released before that.

ConnorDoyle replied:
Updated

vishh commented on pkg/kubelet/cpumanager/util.go:
```diff
@@ -0,0 +1,196 @@
+
+// Returns CPU IDs as a slice sorted by:
+// - socket affinity with result
+// - number of CPUs available on the same sockett
```
nit: `s/sockett/socket`

vishh commented on pkg/kubelet/cpumanager/util.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// Returns CPU IDs as a slice sorted by:
+// - socket affinity with result
```
What is result? This method is complex. It deserves more description on why it's complex. What are the use cases it is trying to tackle?

vishh commented on pkg/kubelet/cpumanager/util.go:
```diff
@@ -0,0 +1,196 @@
+		return nil, fmt.Errorf("not enough cpus available to satisfy request")
+	}
+
+	// Algorithm: topology-aware best-fit
```
With this algorithm, are single sockets filled before other sockets are taken up or will there be spreading?
For apps that run on nodes with static pinning, I assume they will never want to run across sockets. Net performance may be worse than CFS optimizations. How is cross socket assignment prevented?

ConnorDoyle replied:
We will work on improving the user docs and how to avoid cross-socket assignment in practice.

vishh commented on pkg/kubelet/cpumanager/util.go:
```diff
@@ -0,0 +1,196 @@
+limitations under the License.
+*/
+
+package cpumanager
```
This is an important file and it's named `util`. Why not name it `cpu_assigner` or something like that?

ConnorDoyle replied:
Renamed to cpu_assignment.go

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+
+type kletGetter interface {
+	GetPods() []*v1.Pod
+	GetCachedMachineInfo() (*cadvisorapi.MachineInfo, error)
```
In general, I'd like to avoid cross dependencies on kubelet object since it creates circular dependencies.
Can MachineInfo be obtained via the ContainerManager instead? It is expected to own Capacity.

ConnorDoyle replied:
Updated so that ContainerManager passes machine info down to the CPU manager constructor instead.

ConnorDoyle replied:
(in sub-pr)

vishh commented on pkg/api/types.go:
```diff
@@ -3109,6 +3109,8 @@ const (
 	NodeMemoryPressure NodeConditionType = "MemoryPressure"
 	// NodeDiskPressure means the kubelet is under pressure due to insufficient available disk.
 	NodeDiskPressure NodeConditionType = "DiskPressure"
+	// NodeCPUPressure means the kubelet is under pressure due to insufficient available cpu.
```
IIUC, this new condition won't be necessary if kubelet reserves one whole hardware thread by default right?
I suspect thats the default experience most users would want. Reserving that single core would let BE pods run without getting evicted, possibly squeezed on a single core by CFS though. 
CPU pressure as a condition is meant to indicate lack of available CPUs from a usage standpoint. A node under CPU pressure may not be ideal for Bu or BE pods since they will not make progress, or such pods may better be rescheduled.
I'd like to avoid introducing this new condition.

ConnorDoyle replied:
@vishh -- Updated to remove CPU pressure node condition and evictions, require cpu reservation and manage the shared pool accordingly in the static policy.

vishh commented on cmd/kubelet/app/options/options.go:
```diff
@@ -318,6 +318,7 @@ func AddKubeletConfigFlags(fs *pflag.FlagSet, c *kubeletconfig.KubeletConfigurat
 	fs.BoolVar(&c.CgroupsPerQOS, "cgroups-per-qos", c.CgroupsPerQOS, "Enable creation of QoS cgroup hierarchy, if true top level QoS and pod cgroups are created.")
 	fs.StringVar(&c.CgroupDriver, "cgroup-driver", c.CgroupDriver, "Driver that the kubelet uses to manipulate cgroups on the host.  Possible values: 'cgroupfs', 'systemd'")
 	fs.StringVar(&c.CgroupRoot, "cgroup-root", c.CgroupRoot, "Optional root cgroup to use for pods. This is handled by the container runtime on a best effort basis. Default: '', which means use the container runtime default.")
+	fs.StringVar(&c.CPUManagerPolicy, "cpu-manager-policy", c.CPUManagerPolicy, "CPU Manager policy to use. Possible values: 'none', 'static'. Default: 'none'")
```
This is an experimental feature in v1.8 right? The flag needs to indicate that. Also, is there a feature gate?

ConnorDoyle replied:
Added feature gate in the sub-pr.

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	go wait.Until(func() { m.reconcileState() }, time.Second, wait.NeverStop)
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
```
This workflow isn't obvious to me. To help me catch up can you describe (preferably as comments in the PR), when this method is expected to be invoked in the pod lifecycle and why this method has to invoke the UpdateResources method on the CRI, instead of say for example the pod lifecycle handler in kubelet invoking the CRI interfaces?

vishh replied:
Having the overall kubelet interactions documented will be very helpful. TBH, I find it hard to page in interactions between different modules for this feature.

vishh commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+type Manager interface {
+	Start()
+
+	// RegisterContainer registers a container with the cpuset manager
```
Why not have a registration phase at the pod creation time?
For example, kubelet can invoke podContainerManager->EnsureExists(pod) which can ensure that in addition to creating and setting up pod level cgroup, CPU allocations are feasible and even record CPU allocations early on.
If CPU allocations aren't possible (cross socket for example), then the pod can be rejected (this isn't possible yet since EnsureExists is invoked in the sync loop). 
If CPU allocation succeeds, then the podContainerManager can trigger rearrangement of CPU sockets across Bu and BE pods in the background while bringing the node to the desired state. 
This way there isn't a failure possible during container startup. And container restarts do not have to re-register themselves. 
This model will also pave way for identifying devices that are on the same socket as CPUs. For example, get all possible CPU and device allocations and choose the most appropriate combination, then go ahead and reserve CPU and devices.

ConnorDoyle replied:
Sounds like a good plan. However, given the current timeline how important is it to you that this makes it into the first iteration of the feature? We have a number of changes to make based on reviews and would like to get a sense of priority.

vishh replied:
As a community we have not done any code cleanups and reserved time for it. I'm trying to hold the same bar for device plugins too, and there has been some non trivial refactoring happening with the device plugin PR. So ideally, I'd like to hold the same bar for this feature too.

vishh commented on pkg/kubelet/kubelet.go:
```diff
@@ -1103,6 +1123,9 @@ type Kubelet struct {
 	dockerLegacyService dockershim.DockerLegacyService
+
+	// CPU Manager
+	cpuManager cpumanager.Manager
```
I'd prefer not exposing CPU manager to the kubelet. Instead the container manager should be the interface exposed. Kubelet at a high level should deal with Volumes (Storage), Networking (CNI), Resources (Container Manager), Runtime (CRI), Metrics (cAdvisor) APIs only.

ConnorDoyle replied:
This is done in the sub-pr (kubelet wiring)

ConnorDoyle commented on pkg/api/types.go:
```diff
@@ -3109,6 +3109,8 @@ const (
 	NodeMemoryPressure NodeConditionType = "MemoryPressure"
 	// NodeDiskPressure means the kubelet is under pressure due to insufficient available disk.
 	NodeDiskPressure NodeConditionType = "DiskPressure"
+	// NodeCPUPressure means the kubelet is under pressure due to insufficient available cpu.
```
This was necessary because we decided to drive reservations solely by system- and kube-reserved. We considered reserving threads for the shared pool, but it has unfortunate implications for CPU accounting. The current strategy favors pods with a CPU request. A low-water-mark (either by config or unconditional) would favor pods without a CPU request since G/exclusive pods may have to be bounced after scheduling.

ConnorDoyle replied:
Please see the [other comment thread in this PR](https://github.com/kubernetes/kubernetes/pull/49186#discussion_r134920816) where the implications are explained a little more clearly.

ConnorDoyle committed:
Add cpuset helper library.

ConnorDoyle commented:
Rebased

ConnorDoyle committed:
Fix govet and golint problems.

ConnorDoyle committed:
Renamed CPUSet.AsSlice() => CPUSet.ToSlice()

ConnorDoyle committed:
Add CPUSetBuilder, make CPUSet immutable.

ConnorDoyle committed:
Fix golint errors.

rohitsardesai83 commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -116,8 +115,12 @@ func (m *kubeGenericRuntimeManager) startContainer(podSandboxID string, podSandb
 	}
+	err = m.cpuManager.RegisterContainer(pod, container, containerID)
+	if err != nil {
+		m.recorder.Eventf(ref, v1.EventTypeWarning, events.FailedToStartContainer, "Failed to register container with CPU manager: %v", err)
```
Should we introduce a new events constant here say **events.FailedToRegisterContainer** or something ?
since this is not a failure to start the container but a registration failure ?

ConnorDoyle replied:
Not sure, haven't examined the event classifications closely enough to have an opinion about this. However, we are bailing out of `startContainer` on account of the registration failure so it doesn't seem too far off.

rohitsardesai83 commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -810,6 +813,7 @@ func (m *kubeGenericRuntimeManager) removeContainer(containerID string) error {
 	if err := m.removeContainerLog(containerID); err != nil {
 		return err
 	}
+	m.cpuManager.UnregisterContainer(containerID)
```
what do we do in case we get an error while unregistering the container ?

ConnorDoyle replied:
This has been updated so that errors are surfaced to the caller of killContainer and removeContainer.

ConnorDoyle commented on pkg/kubelet/cpumanager/util.go:
```diff
@@ -0,0 +1,196 @@
+		return nil, fmt.Errorf("not enough cpus available to satisfy request")
+	}
+
+	// Algorithm: topology-aware best-fit
```
This was discussed as part of the design proposal [here](https://github.com/kubernetes/community/pull/654#discussion_r120685990). TL;DR: Socket and physical core assignment is delivered on a best-effort basis by the static policy.

vishh replied:
Can the documentation be cross-linked?

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	go wait.Until(func() { m.reconcileState() }, time.Second, wait.NeverStop)
+}
+
+func (m *manager) RegisterContainer(p *v1.Pod, c *v1.Container, containerID string) error {
```
We can definitely explore moving the call to UpdateContainerResources to the lifecycle handlers.

derekwaynecarr replied:
this flow was not obvious to me, but i have a better understanding of the intent.

my high-level understanding:

kubelet syncPod
- kubelet.go knows about containerManager only
- it asks a containerManager for a "PodContainerManager"
- it asks pod container manager to verify a pod cgroup exists (if needed)
- kubelet then calls containerruntime.SyncPod(...)
- container runtime sync pod creates container
- container runtime sync pod then registers with cpu manager
- cpu manager then calls back to container runtime and updates container resources with new cpu set assignment
- container runtime then starts container

let me think on a better flow on this over night, but i understand that this may be non-obvious.

derekwaynecarr replied:
the major question in my head is if the container runtime SyncPod operation should take a ContainerResourceManager as an argument.  The ContainerResourceManager would be created from the PodContainerManager in the main syncPod loop.  In this way, kubelet.go at a top-level just knows about ContainerManager.

It would have an operation for "PreStartContainer" and "PostStopContainer" which would match the intent with "RegisterContainer" and "UnregisterContainer".  I think that would potentially be clearer.

@vishh -- does that align with your thoughts?

vishh replied:
+1. I do not mind re-using PodContainerManger since kubelet mostly manages pods.

ConnorDoyle commented on pkg/kubelet/kubelet.go:
```diff
@@ -1103,6 +1123,9 @@ type Kubelet struct {
 	dockerLegacyService dockershim.DockerLegacyService
+
+	// CPU Manager
+	cpuManager cpumanager.Manager
```
:+1: Currently investigating burying CPU manager inside container manager.

ConnorDoyle commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -810,6 +813,7 @@ func (m *kubeGenericRuntimeManager) removeContainer(containerID string) error {
 	if err := m.removeContainerLog(containerID); err != nil {
 		return err
 	}
+	m.cpuManager.UnregisterContainer(containerID)
```
Right now the error is discarded but that probably is not the right thing to do. In case unregistration fails we probably want to fail `removeContainer` so that it gets called again. Otherwise we could leak/strand CPUs.

CC @sjenning

ConnorDoyle committed:
Addressed some review comments.

- Typos
- Type naming.
- File rename.
- Added alpha warning to --cpu-manager-policy config description.

ConnorDoyle commented on cmd/kubelet/app/options/options.go:
```diff
@@ -318,6 +318,7 @@ func AddKubeletConfigFlags(fs *pflag.FlagSet, c *kubeletconfig.KubeletConfigurat
 	fs.BoolVar(&c.CgroupsPerQOS, "cgroups-per-qos", c.CgroupsPerQOS, "Enable creation of QoS cgroup hierarchy, if true top level QoS and pod cgroups are created.")
 	fs.StringVar(&c.CgroupDriver, "cgroup-driver", c.CgroupDriver, "Driver that the kubelet uses to manipulate cgroups on the host.  Possible values: 'cgroupfs', 'systemd'")
 	fs.StringVar(&c.CgroupRoot, "cgroup-root", c.CgroupRoot, "Optional root cgroup to use for pods. This is handled by the container runtime on a best effort basis. Default: '', which means use the container runtime default.")
+	fs.StringVar(&c.CPUManagerPolicy, "cpu-manager-policy", c.CPUManagerPolicy, "CPU Manager policy to use. Possible values: 'none', 'static'. Default: 'none'")
```
Added the alpha warning, but we do still need the feature gate. There is a note about this in the PR breakdown plan in the PR description.

ConnorDoyle commented on pkg/kubelet/cpumanager/policy_none.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (p *nonePolicy) UnregisterContainer(s state.State, containerID string) error {
+	glog.Infof("[cpumanager] none policy: UnregisterContainer [%s]", containerID)
```
Removed

ConnorDoyle commented on pkg/kubelet/cpumanager/policy_none.go:
```diff
@@ -0,0 +1,57 @@
+}
+
+func (p *nonePolicy) RegisterContainer(s state.State, pod *v1.Pod, container *v1.Container, containerID string) error {
+	glog.Infof("[cpumanager] none policy: RegisterContainer [%s] (%s) of pod %s", containerID, container.Name, pod.Name)
```
Removed

ConnorDoyle commented:
Addressed some of the review comments. The following TODOs are not done yet but are hidden by the GitHub UI:

- [ ] [comment from @vishh](https://github.com/kubernetes/kubernetes/pull/49186#discussion_r134382102) _What is result? This method is complex. It deserves more description on why it's complex. What are the use cases it is trying to tackle?_
- [ ] [comment from @vishh](https://github.com/kubernetes/kubernetes/pull/49186#discussion_r134382283) _How is cross-socket assignment prevented?_
- [ ] [comment from @vishh](https://github.com/kubernetes/kubernetes/pull/49186#discussion_r134384168) _Why not have a registration phase at the pod creation time? ..._
- [ ] [comment from @vishh](https://github.com/kubernetes/kubernetes/pull/49186#pullrequestreview-57666570) _Add a feature gate..._

ConnorDoyle committed:
Moved cpuset to /pkg/kubelet/cm.

ConnorDoyle committed:
Release lock before UpdateContainerResources.

ConnorDoyle committed:
Removed eviction and node condition from CPU man.

- Require kubereserved.cpu + systemreserved.cpu > 0.
- Updated shared pool and exclusive allocation accordingly.

vishh commented:
Awesome. Thanks @ConnorDoyle!

On Fri, Aug 25, 2017 at 11:01 AM, Connor Doyle <notifications@github.com>
wrote:

> *@ConnorDoyle* commented on this pull request.
> ------------------------------
>
> In pkg/api/types.go
> <https://github.com/kubernetes/kubernetes/pull/49186#discussion_r135318420>
> :
>
> > @@ -3109,6 +3109,8 @@ const (
>  	NodeMemoryPressure NodeConditionType = "MemoryPressure"
>  	// NodeDiskPressure means the kubelet is under pressure due to insufficient available disk.
>  	NodeDiskPressure NodeConditionType = "DiskPressure"
> +	// NodeCPUPressure means the kubelet is under pressure due to insufficient available cpu.
>
> @vishh <https://github.com/vishh> -- Updated to remove CPU pressure node
> condition and evictions, require cpu reservation and manage the shared pool
> accordingly in the static policy.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/49186#discussion_r135318420>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKMNXpgyzhP9AjFQ8PBPRWWQ-LG70ks5sbwvpgaJpZM4OcVm->
> .
>

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	for _, container := range status.ContainerStatuses {
+		if container.Name == name && container.ContainerID != "" {
+			// hack hack strip docker:// hack hack
+			return container.ContainerID[9:], nil
```
yes, updated this in https://github.com/kubernetes/kubernetes/pull/51357/files#diff-67c46e7cbb41ea5e675000a3298e6b32R225

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	// UnregisterContainer is called near UnregisterContainer() so that the manager
+	// stops trying to update that container in the reconcilation loop and
+	// any CPUs dedicated to the container are freed to the shared pool.
+	UnregisterContainer(containerID string) error
```
It can error whenever the underlying policy errors (we don't know yet all of those cases.) Updated the call sites to propagate the errors.

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+	}
+
+	for _, pod := range m.kletGetter.GetPods() {
+		for _, container := range pod.Spec.Containers {
```
Updated in sub-PR.

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+		return err
+	}
+	cpuset := m.state.GetCPUSetOrDefault(containerID)
+	err = m.containerRuntime.UpdateContainerResources(
```
@sjenning wdyt about this? Does it make more sense to try to dig up the full set of LinuxContainerResources or perhaps rename this function to sth like `PatchContainerResources`?

ConnorDoyle commented on pkg/kubelet/cpumanager/cpu_manager.go:
```diff
@@ -0,0 +1,363 @@
+			}
+
+			glog.Infof("[cpumanager] reconcileState: updating container (pod: %s, container: %s, container id: %s, cpuset: \"%v\")", pod.Name, container.Name, containerID, cset)
+			err = m.containerRuntime.UpdateContainerResources(
```
@sjenning wdyt about this? Does it make more sense to try to dig up the full set of LinuxContainerResources or perhaps rename this function to sth like PatchContainerResources?

Is resolving the usage of the new API a blocker for this feature?

ConnorDoyle commented:
Addressed the vast majority of review comments and broke this down into smaller PRs. Those are all linked from the top PR description comment here. Continuing reviews together with @derekwaynecarr there. Thanks everyone for reviews on this monster PR -- closing.
